sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala

```diff
@@ -47,15 +48,26 @@ object CommandUtils extends Logging {
     }
   }
 
-  def calculateTotalSize(sessionState: SessionState, catalogTable: CatalogTable): BigInt = {
+    def calculateTotalSize(spark: SparkSession, catalogTable: CatalogTable): BigInt = {
+
+    val sessionState = spark.sessionState
+    val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
+
     if (catalogTable.partitionColumnNames.isEmpty) {
-      calculateLocationSize(sessionState, catalogTable.identifier, catalogTable.storage.locationUri)
+      calculateLocationSize(sessionState, catalogTable.identifier,
+          catalogTable.storage.locationUri)
     } else {
       // Calculate table size as a sum of the visible partitions. See SPARK-21079
       val partitions = sessionState.catalog.listPartitions(catalogTable.identifier)
-      partitions.map { p =>
-        calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-      }.sum
+      val paths = partitions.map(x => new Path(x.storage.locationUri.get.getPath))
+      val pathFilter = new PathFilter {
+        override def accept(path: Path): Boolean = {
+          !path.getName.startsWith(stagingDir)
+        }
+      }
+      val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(paths,
+        sessionState.newHadoopConf(), pathFilter, spark).flatMap(x => x._2)
```

Comment:
The tests are passing but this line is incorrect. 

@gatorsmile @maropu, PathFilter is not serializable, How do we pass PathFilter in a serializable manner? I checked the code and one other way is to use `FileInputFormat.getInputPathFilter`/`FileInputFormat.setInputPathFilter` but I couldn't get it to work. 

Also, is it okay if we filter the `Seq[(Path, Seq[FileStatus])]` returned by `bulkListLeafFiles` and remove `stagingDir` files?

Reply:
The above approach might not work too. In the earlier implementation there was a check from recursively listing files from certain directories (`stagingDir`) and having a pathFilter might not be the right approach. 

So I wanted to introduce a list of strings called `filterDir` as a new parameter to `bulkListLeafFiles` which can be used to check if a particular directory can be recursed further.

Reply:
```Scala
class PathFilterIgnoreNonData(stagingDir: String) extends PathFilter with Serializable {
  override def accept(path: Path): Boolean = {
    val fileName = path.getName
    (!fileName.startsWith(stagingDir) &&
      // Ignore metadata files starting with "_" (for example, files created by
      // DirectoryAtomicCommitProtocol) when computing the location size
      !fileName.startsWith("_"))
  }
}
```

Reply:
Thank you, I have made the changes. Can you review this?